### PR TITLE
Vue JS streams rewrite part 2

### PR DESCRIPTION
**Pull request by christophehenry**
* Fix placement of "new posts label, should be horizontally centered, just below navbar
* Maintains list of ID's for new content waiting on server.
* Content timestamp is a link to the "single content view"
* End of replies section has a "reply" button
* Clicking reply button goes to reply create page

### CODE REVIEW

File: socialhome/streams/app/components/Stream.vue [hunk 1]
```diff
@@ -1,45 +1,47 @@
...
+                    :current-browsing-profile-id="$store.state.globalStore.currentBrowsingProfileId"
+                />
             </div>
         </div>
     </div>
 </template>
 
 <script>
 import Vue from "vue"
+import {mapState} from "vuex"
```
Comment (christophehenry) [inline socialhome/streams/app/components/Stream.vue]: Uneeded.

File: socialhome/streams/app/components/StreamElement.vue [hunk 2]
```diff
@@ -6,13 +6,11 @@
...
             :is-user-author="isUserAuthor"
             :is-user-local="isUserLocal"
-            :current-browsing-profile-id="currentBrowsingProfileId"
-            :is-user-authenticated="isUserAuthenticated"
         >
         </author-bar>
         <div class="grid-item-bar d-flex justify-content-start">
             <div class="mt-1">
-                <a :ref="contentUrl" :title="timestamp">
+                <a :href="contentUrl" :title="timestamp">
```
Comment (christophehenry) [inline socialhome/streams/app/components/StreamElement.vue]: `b-button`

File: socialhome/streams/app/tests/components/AuthorBar.tests.js [hunk 3]
```diff
@@ -7,6 +7,7 @@ import Vue from "vue"
 import VueMasonryPlugin from "vue-masonry"
 
 import AuthorBar from "streams/app/components/AuthorBar.vue"
+import globalStore from "streams/app/stores/globalStore"
```
Comment (christophehenry) [inline socialhome/streams/app/tests/components/AuthorBar.tests.js]: Unneeded

File: socialhome/streams/app/tests/fixtures/AuthorBar.fixtures.js [hunk 4]
```diff
@@ -1,19 +1,18 @@
...
 const getAuthorBarPropsData = function (args = {}) {
     return _.defaults({}, args, {
-        handle: "",
-        name: "",
-        guid: "",
-        currentBrowsingProfileId: "",
-        homeUrl: "",
-        absoluteUrl: "",
-        imageUrlSmall: "",
+        handle: faker.internet.email(),
```
Comment (jaywink) [inline socialhome/streams/app/tests/fixtures/AuthorBar.fixtures.js]: Use `faker.internet.exampleEmail()` to make a safe email that will not accidentally deliver anywhere.

File: socialhome/streams/app/tests/fixtures/AuthorBar.fixtures.js [hunk 5]
```diff
@@ -1,19 +1,18 @@
...
+        name: faker.name.findName(),
+        guid: `${faker.random.number()}`,
```
Comment (jaywink) [inline socialhome/streams/app/tests/fixtures/AuthorBar.fixtures.js]: `faker.random.uuid()` is a real fake guid.

File: socialhome/streams/app/tests/fixtures/AuthorBar.fixtures.js [hunk 6]
```diff
@@ -1,19 +1,18 @@
...
+        homeUrl: "https://127.0.0.1",
+        absoluteUrl: "https://127.0.0.1",
+        imageUrlSmall: "https://127.0.0.1",
```
Comment (jaywink) [inline socialhome/streams/app/tests/fixtures/AuthorBar.fixtures.js]: `https://127.0.0.1/image.png` would be better.

File: socialhome/streams/app/tests/fixtures/StreamElementReactionsBar.fixtures.js [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+import faker from "faker"
+
+import {getAuthorBarPropsData} from "streams/app/tests/fixtures/AuthorBar.fixtures";
+
+
+const getStreamElementReactionsBarPropsData = function (args = {}) {
+    return _.defaults({}, args, {
+        id: faker.random.number(),
+        repliesCount: faker.random.number(),
+        sharesCount: faker.random.number(),
```
Comment (jaywink) [inline socialhome/streams/app/tests/fixtures/StreamElementReactionsBar.fixtures.js]: `hasShared - boolean` missing?

File: socialhome/streams/app/tests/fixtures/StreamElementReactionsBar.fixtures.js [hunk 8]
```diff
@@ -0,0 +1,15 @@
+import _ from "lodash/core"
+import faker from "faker"
+
+import {getAuthorBarPropsData} from "streams/app/tests/fixtures/AuthorBar.fixtures";
```
Comment (jaywink) [inline socialhome/streams/app/tests/fixtures/StreamElementReactionsBar.fixtures.js]: This import seems unnecessary.

**Review by jaywink — CHANGES_REQUESTED** [review]

File: socialhome/streams/app/stores/streamStore.state.js [hunk 9]
```diff
@@ -1,24 +1,14 @@
...
-        stream: {
-            hasNewContent: false,
-            newContentLengh: 0,
-            streamName: _get(window, ["context", "streamName"], ""),
-        },
-        isUserAuthenticated: _get(window, ["context", "isUserAuthenticated"], false),
-        showAuthorBar: _get(window, ["context", "showAuthorBar"], false),
-        currentBrowsingProfileId: (_isNumber(_get(window, ["context", "currentBrowsingProfileId"], undefined))
-            ? `${window.context.currentBrowsingProfileId}` : undefined),
+        contentList: _keyBy(_get(window, ["context", "contentList"], []), item => item.id),
```
[the remark this answers is not in the document]
Comment (jaywink) [reply socialhome/streams/app/stores/streamStore.state.js]: Btw, I hope this doesn't mess with the order since the stream must be rendered in the correct order, not by id.

File: socialhome/streams/app/components/StreamElement.vue [hunk 10]
```diff
@@ -67,66 +49,21 @@ export default Vue.component("stream-element", {
...
-            }
-            this.$http.delete(`/api/content/${this.id}/share/`)
-                .then(() => {
-                    this.$data.showSharesBox = false
-                    this.sharesCount$ -= 1
-                    this.hasShared$ = false
-                })
-                .catch(err => console.error(err) /* TODO: Proper error handling */)
-        },
+        deleteUrl(){
```
[the remark this answers is not in the document]
Comment (christophehenry) [reply socialhome/streams/app/components/StreamElement.vue]: It's planned. It's just I fixed the delete button which wasn't working.

File: socialhome/streams/app/components/StreamElement.vue [hunk 11]
```diff
@@ -6,59 +6,41 @@
...
         <div class="grid-item-bar d-flex justify-content-start">
             <div class="mt-1">
-                <a :ref="contentUrl" :title="timestamp">
+                <b-button :href="contentUrl" :title="timestamp" variant="link">
                     {{ humanizedTimestamp }}<span v-if="edited"> {{ editedText }}</span>
-                </a>
+                </b-button>
                 <div v-if="isUserAuthor">
-                    <a :href="updateUrl">
+                    <b-button :href="updateUrl">
```
[the remark this answers is not in the document]
Comment (christophehenry) [reply socialhome/streams/app/components/StreamElement.vue]: `b-button` generates a link used with attribute `href`. I prefer to use Bootstrap components as much as possible. I'm so concerned about how things can go wrong with CSS.

It does not look the same as before yet. But I need to change the HTML a bit. I broke the design.

File: socialhome/streams/app/stores/globalStore.js [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+            },
+            isUserAuthenticated: _get(window, ["context", "isUserAuthenticated"], false),
+            currentBrowsingProfileId: (_isNumber(_get(window, ["context", "currentBrowsingProfileId"], undefined))
+                ? `${window.context.currentBrowsingProfileId}` : undefined),
+        },
+    })
+}
+
+const globalStore = newGlobalStore()
+export default globalStore
```
[the remark this answers is not in the document]
Comment (christophehenry) [reply socialhome/streams/app/stores/globalStore.js]: `appStore` is fine. I used the term `global` because it's actually a singleton in the code.

File: socialhome/streams/app/stores/streamStore.js [hunk 13]
```diff
@@ -1,25 +1,31 @@
...
 
 Vue.use(Vuex)
 
-function newinstance(WebSocketImpl = ReconnectingWebSocket) {
+function newStreamStore(options) {
     const state = getState()
-    const store = new Vuex.Store({state, mutations, actions})
+    const opts = _defaults({}, {state, mutations, actions}, options)
+    const WebSocketImpl = _get(opts, ["WebSocketImpl"], ReconnectingWebSocket)
+    delete opts.WebSocketImpl
```
[the remark this answers is not in the document]
Comment (christophehenry) [reply socialhome/streams/app/stores/streamStore.js]: Test purposes. It has no other use whatsoever. I didn't find a correct way to spy on websocket implementation except by injecting it. After that, I delete it to prevent it from being added to the store's properties. Which is a bit dumb now that I think about it, because it's not used outside tests. I suppose I should add a comment.

File: socialhome/streams/app/stores/streamStore.state.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (christophehenry) [reply socialhome/streams/app/stores/streamStore.state.js]: > Btw, I hope this doesn't mess with the order since the stream must be rendered in the correct order, not by id.

Good question. Object are not ordered in JS. So my implementation is wrong. I have to use [Maps](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map).

File: socialhome/streams/app/components/StreamElement.vue (hunk 11 above)
[the remark this answers is not in the document]
Comment (jaywink) [reply socialhome/streams/app/components/StreamElement.vue]: My point is to me it seems 1) wrong to use a button when you want a link and 2) using a `a` tag will not require any extra styling. So by using the wrong element, we need to add extra directives and/or style to make it look and behave like a normal link.

Additionally, browsers might not interpret a button the same way as a link. For this reason, I think links should be kept as standard HTML specification links, not buttons.

[the remark this answers is not in the document]
Comment (christophehenry) [reply socialhome/streams/app/components/StreamElement.vue]: > My point is to me it seems 1) wrong to use a button when you want a link

It's not a button. If you add `href` attribute to `b-button`, it generates a link.

File: socialhome/streams/app/stores/streamStore.state.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (christophehenry) [reply socialhome/streams/app/stores/streamStore.state.js]: Ah, crap. Maps won't work either. Vue can only watch native objects.

File: socialhome/streams/app/components/StreamElement.vue (hunk 11 above)
[the remark this answers is not in the document]
Comment (jaywink) [reply socialhome/streams/app/components/StreamElement.vue]: That is like having a class called `Profile` but creating an instance of `Webpage` when the user gives a `url` argument to it. I mean, wtf ;)

Well, I wont fight it, though I think it doesn't make any sense at all to use `b-button` if a normal `a` is *what we really want*.

File: socialhome/streams/app/stores/streamStore.state.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (jaywink) [reply socialhome/streams/app/stores/streamStore.state.js]: What about storing the list and the objects separately, like;

    contentList: [1,2,3],
    contents: {1: {}, 2:{}, 3:{}},

This way 1) order is guaranteed and 2) objects can be found easily from the components, ensuring we don't need to duplicate store into components as mutable local state.

This allows also inserting directly to the `contentList` from a websocket `new` message. When the object itself appears in `contents`, it will then render it.

File: socialhome/streams/app/stores/streamStore.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (jaywink) [reply socialhome/streams/app/stores/streamStore.js]: Generally of course it would be nice to avoid code that exists only for tests, but if you didn't find any other way to do it then a clear comment please and let's live with it - can always fix it if another way comes up 👍

File: socialhome/streams/app/components/StreamElement.vue (hunk 11 above)
[the remark this answers is not in the document]
Comment (christophehenry) [reply socialhome/streams/app/components/StreamElement.vue]: Ok, I went back to an `a`. Using Bootstrap classes modifies the design and design is not the goal here.

File: socialhome/streams/app/components/StreamElement.vue (hunk 10 above)
[the remark this answers is not in the document]
Comment (christophehenry) [reply socialhome/streams/app/components/StreamElement.vue]: `contentUrl` [is a bit more complex](https://github.com/jaywink/socialhome/blob/master/socialhome/content/models.py#L154). Not sure it's a good idea.

File: socialhome/streams/app/components/StreamElementReactionsBar.vue [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+            return this.$store.state.isUserAuthenticated || this.sharesCount$ > 0
+        },
+        replyUrl(){
+            return Urls["content:reply"]({pk: this.id})
+        }
+    },
+    methods: {
+        expandShares() {
+            this.showSharesBox = !this.showSharesBox
+            this.showRepliesBox = this.showSharesBox ? false : this.showRepliesBox
```
[the remark this answers is not in the document]
Comment (christophehenry) [reply socialhome/streams/app/components/StreamElementReactionsBar.vue]: My own initative. I though it's best to show only one button at a time. It simplifies the UI.

File: socialhome/streams/app/components/Stream.vue [hunk 15]
```diff
@@ -1,30 +1,27 @@
 <template>
     <div>
         <div class="container-flex">
-            <div v-show="$store.state.stream.hasNewContent" class="new-content-container">
+            <div v-show="$store.state.hasNewContent" class="new-content-container">
```
Comment (jaywink) [inline socialhome/streams/app/components/Stream.vue:4]: Change this class to an `id` and you don't have to duplicate the CSS. Currently in the non-Vue version this is an `id`.

File: socialhome/static/sass/streams.scss [hunk 16]
```diff
@@ -1,7 +1,8 @@
 @import "variables";
 
 /* New content container */
-#new-content-container {
+#new-content-container,
+.new-content-container {
```
Comment (jaywink) [inline socialhome/static/sass/streams.scss:5]: These duplications in this file are unnecessary as other comment re class vs id.

File: socialhome/streams/app/components/ReactionsBar.vue [hunk 17]
```diff
@@ -0,0 +1,106 @@
+<template>
+    <div class="grid-item-bar d-flex justify-content-start">
+        <slot />
```
Comment (jaywink) [inline socialhome/streams/app/components/ReactionsBar.vue:3]: What is this? Does it receive the stuff passed into as content when calling this component?

File: socialhome/streams/app/components/StreamElement.vue [hunk 18]
```diff
@@ -6,126 +6,66 @@
...
-        <div class="grid-item-bar d-flex justify-content-start">
+        <reactions-bar
+            :id="id"
+            :replies-count="repliesCount"
+            :shares-count="sharesCount"
+            :has-shared="hasShared"
+        >
             <div class="mt-1">
-                <a :ref="contentUrl" :title="timestamp">
+                <a :href="contentUrl" :title="timestamp" variant="link" class="unstyled-link">
```
Comment (jaywink) [inline socialhome/streams/app/components/StreamElement.vue]: It's an `a` tag, does it need `variant`?

File: socialhome/streams/app/components/StreamElement.vue [hunk 19]
```diff
@@ -6,126 +6,66 @@
             v-bind="author"
             :is-user-author="isUserAuthor"
             :is-user-local="isUserLocal"
-            :current-browsing-profile-id="currentBrowsingProfileId"
-            :is-user-authenticated="isUserAuthenticated"
         >
         </author-bar>
-        <div class="grid-item-bar d-flex justify-content-start">
+        <reactions-bar
```
Comment (jaywink) [inline socialhome/streams/app/components/StreamElement.vue:11]: How come we're calling a component and then still having content here? What does it mean?

File: socialhome/streams/app/components/StreamElement.vue [hunk 20]
```diff
@@ -6,126 +6,66 @@
...
+        >
             <div class="mt-1">
-                <a :ref="contentUrl" :title="timestamp">
+                <a :href="contentUrl" :title="timestamp" variant="link" class="unstyled-link">
                     {{ humanizedTimestamp }}<span v-if="edited"> {{ editedText }}</span>
                 </a>
-                <div v-if="isUserAuthor">
-                    <a :href="updateUrl">
+                &nbsp;
+                <template v-if="isUserAuthor">
```
Comment (jaywink) [inline socialhome/streams/app/components/StreamElement.vue]: What does the `template` tag do here, why that instead of `div`?

File: socialhome/streams/app/components/StreamElement.vue [hunk 21]
```diff
@@ -6,126 +6,66 @@
...
+                    <a :href="updateUrl" variant="link">
```
Comment (jaywink) [inline socialhome/streams/app/components/StreamElement.vue]: It's an `a` tag, does it need `variant`?

File: socialhome/streams/app/components/StreamElement.vue [hunk 22]
```diff
@@ -6,126 +6,66 @@
...
                         <i class="fa fa-pencil" title="Update" aria-label="Update"></i>
                     </a>
-                    <a :href="deleteUrl">
+                    &nbsp;
+                    <a :href="deleteUrl" variant="link">
```
Comment (jaywink) [inline socialhome/streams/app/components/StreamElement.vue]: It's an `a` tag, does it need `variant`?

File: socialhome/streams/app/components/ReactionsBar.vue [hunk 23]
```diff
@@ -0,0 +1,106 @@
...
+        <div class="ml-auto grid-item-reactions mt-1">
+            <b-button
```
Comment (jaywink) [inline socialhome/streams/app/components/ReactionsBar.vue:5]: Does the design match the old one? This used to be a `div`.

selection_415

File: socialhome/streams/app/components/ReactionsBar.vue [hunk 24]
```diff
@@ -0,0 +1,106 @@
...
+                v-if="showShares"
+                :class="{'item-reaction-shared': hasShared$, 'item-reaction': true}"
```
Comment (jaywink) [inline socialhome/streams/app/components/ReactionsBar.vue]: I believe you can do `class="item-reaction" :class="{'item-reaction-shared': hasShared$}"` .. ie both `class` and `:class` can exist without having to make it an "if" statement by giving an extra `true` value.

File: socialhome/streams/app/components/ReactionsBar.vue [hunk 25]
```diff
@@ -0,0 +1,106 @@
...
+                    <span class="item-reaction-counter">{{ repliesCount$ }}</span>
+                </span>
+            </b-button>
+        </div>
+        <div v-if="showSharesBox" class="content-actions">
+            <b-button v-if="hasShared$" variant="secondary" @click.prevent.stop="unshare">Unshare</b-button>
+            <b-button v-if="!hasShared$" variant="secondary" @click.prevent.stop="share">Share</b-button>
+        </div>
+        <div v-if="showRepliesBox">
+            <div class="replies-container"></div>
```
Comment (jaywink) [inline socialhome/streams/app/components/ReactionsBar.vue]: This should be outside `<div v-if="showRepliesBox">` before it.

File: socialhome/streams/app/components/ReactionsBar.vue [hunk 26]
```diff
@@ -0,0 +1,106 @@
...
+            return this.$store.state.isUserAuthenticated || this.sharesCount$ > 0
+        },
+        replyUrl() {
+            return Urls["content:reply"]({pk: this.id})
+        },
+    },
+    methods: {
+        expandShares() {
+            this.showSharesBox = !this.showSharesBox
+            this.showRepliesBox = this.showSharesBox ? false : this.showRepliesBox
```
Comment (jaywink) [inline socialhome/streams/app/components/ReactionsBar.vue]: I don't think this is a good idea. Reason: replies. If there are replies, say 10 of them, clicking the share shouldn't make the replies disappear. And if it doesn't, it really doesn't make sense to make the reply button *at the end of the replies*  to disappear.

I would prefer to keep old functionality where the reaction icons just act as toggles and don't depend on each other. It would be different if they were somehow dependent on each other in concept (like in a "you can do one or the other"), but they're not, they're completely different things.

File: socialhome/streams/app/components/ReactionsBar.vue [hunk 27]
```diff
@@ -0,0 +1,106 @@
...
+            }
+            this.$http.delete(`/api/content/${this.id}/share/`)
+                .then(() => {
+                    this.showSharesBox = false
+                    this.sharesCount$ -= 1
+                    this.hasShared$ = false
+                })
+                .catch(err => console.error(err) /* TODO: Proper error handling */)
+        },
+        expandComments() {
```
Comment (jaywink) [inline socialhome/streams/app/components/ReactionsBar.vue]: One mistake I've made with the Python files - not organizing methods alphabetically. Let's not do that mistake here :)

File: socialhome/streams/views.py [hunk 28]
```diff
@@ -48,14 +48,11 @@ def dump_content(content):
                 "timestamp": content.formatted_timestamp,
                 "humanizedTimestamp": content.humanized_timestamp,
                 "edited": bool(content.edited),
-                "childrenCount": content.reply_count,
+                "repliesCount": content.children_count,
```
Comment (jaywink) [inline socialhome/streams/views.py]: It's now `content.reply_count` in master 👍

File: socialhome/streams/app/stores/streamStore.operations.js [hunk 29]
```diff
@@ -1,29 +1,46 @@
-const stateOperations = {
+import _forEach from "lodash/forEach"
+
+
+const streamStoreOerations = {
```
Comment (jaywink) [inline socialhome/streams/app/stores/streamStore.operations.js]: Typo, should be `streamStoreOperations`.

File: socialhome/streams/app/components/Stream.vue [hunk 30]
```diff
@@ -1,30 +1,27 @@
...
                 <div class="stamped">
-                    <stamped-element></stamped-element>
+                    <stamped-element />
                 </div>
                 <div class="grid-sizer"></div>
                 <div class="gutter-sizer"></div>
                 <stream-element
                     class="grid-item"
-                    v-for="content in $store.state.contentList"
+                    v-for="content in $store.getters.contentList"
```
Comment (jaywink) [inline socialhome/streams/app/components/Stream.vue:19]: I'm wondering, won't any change in any content always re-render every single content instead of one of them? Any idea? Also does Vue have to constantly check the getter whether there are new items in the list? I really have no idea how it should work, just wondering out aloud.

We need only one content to re-render when it changes or new appears, otherwise we don't get any benefit from using Vue. Currently the stream re-renders everything on one change.

**Review by jaywink — CHANGES_REQUESTED** [review]
A few change requests and comments, some because I don't know what is going on and the PR is getting to be quite big and it's getting hard to follow. Hope we can clean up and merge soon 👍

File: socialhome/streams/app/components/Stream.vue (hunk 15 above)
Comment (christophehenry) [reply socialhome/streams/app/components/Stream.vue:4]: There are two main reasons for using classes here:

1. classes are more efficient than ids for applying CSS (see for instance answer [here](https://stackoverflow.com/questions/7623947/css-id-class-vs-class-performance-which-is-better#7623970)),
2. many of the ids you created are bound to existant JS code, which interfere with Vue rewriting.

Comment (jaywink) [reply socialhome/streams/app/components/Stream.vue:4]: Ah right, the existing JS 👍

File: socialhome/streams/app/components/Stream.vue (hunk 30 above)
Comment (christophehenry) [reply socialhome/streams/app/components/Stream.vue:19]: According to the documentation Vuex' getters are similar to component's computed properties which are cached. As long as the base data don't change, the list don't change. But I suppose the list is entirely recomputed when an element is inserted to or removed from the original list. Which can be heavy, yes.

Good thing is that everything is async in Vue. The view is rerendered the next tick after the property was entirely recomputed so that shouldn't slower the stream in any way.

As for the view re-render, Vue only renders what changed thanks to virtual DOM. This is why list rendering requires [the `id` atribute](https://github.com/jaywink/socialhome/pull/273/files/c8a6be58b6883dfafa3b492096834bf3e1925a25#diff-f6bf769c688863e153cb8f8d589a3803R22), so that only new or removed elements change.

[There's an example close to that in Vuex documentation](https://vuex.vuejs.org/en/getters.html). You can read about the in-place patch strategy [here](https://vuejs.org/v2/guide/list.html#key).

We may face problems with the child component's internal state, though. But as you noticed, our strategy is far from ideal, though I'd consider it out of scope for this PR. I don't think it's time for early optimisation. Especially since we are still discovering Vue's good practices.

File: socialhome/streams/app/components/ReactionsBar.vue (hunk 17 above)
Comment (christophehenry) [reply socialhome/streams/app/components/ReactionsBar.vue:3]: Yes. It's mainly to keep the current design while leaving the timestamp, delete and edit links, outside the component. It's not pretty, but I think this will change when I start to work on the design.

File: socialhome/streams/app/components/ReactionsBar.vue (hunk 23 above)
Comment (christophehenry) [reply socialhome/streams/app/components/ReactionsBar.vue:5]: Yes, it does. Styling is attached to `.item-reaction` class.

File: socialhome/streams/app/components/ReactionsBar.vue (hunk 24 above)
Comment (christophehenry) [reply socialhome/streams/app/components/ReactionsBar.vue]: I don't think so but it seems `:class="[{'item-reaction-shared': hasShared$},'item-reaction']"` will work.

File: socialhome/streams/app/components/ReactionsBar.vue (hunk 26 above)
Comment (christophehenry) [reply socialhome/streams/app/components/ReactionsBar.vue]: Alright.

File: socialhome/streams/app/components/StreamElement.vue (hunk 18 above)
Comment (christophehenry) [reply socialhome/streams/app/components/StreamElement.vue]: No, it does not. It's... It's... It's a... typo...

File: socialhome/streams/app/components/StreamElement.vue (hunk 19 above)
Comment (christophehenry) [reply socialhome/streams/app/components/StreamElement.vue:11]: Content inserted inside the component tag can be rendered by the component with [the `<slot />` tag](https://vuejs.org/v2/guide/components.html#Single-Slot).

File: socialhome/streams/app/components/StreamElement.vue (hunk 20 above)
Comment (christophehenry) [reply socialhome/streams/app/components/StreamElement.vue]: The `template` tag will not be rendered. Only what it contains. It's usefull to do [condiftionnal groups](https://vuejs.org/v2/guide/conditional.html#Conditional-Groups-with-v-if-on-lt-template-gt).

File: socialhome/streams/app/components/StreamElement.vue (hunk 21 above)
Comment (christophehenry) [reply socialhome/streams/app/components/StreamElement.vue]: # TYPO ALEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEERT!

File: socialhome/streams/app/components/StreamElement.vue (hunk 22 above)
Comment (christophehenry) [reply socialhome/streams/app/components/StreamElement.vue]: Does it look like a typo? Hmm... :D

File: socialhome/streams/views.py (hunk 28 above)
Comment (christophehenry) [reply socialhome/streams/views.py]: Breaking changes are evil! :D

File: socialhome/streams/app/stores/streamStore.operations.js (hunk 29 above)
Comment (christophehenry) [reply socialhome/streams/app/stores/streamStore.operations.js]: Hmm... What's a typo? :D

File: socialhome/streams/app/components/ReactionsBar.vue (hunk 24 above)
Comment (jaywink) [reply socialhome/streams/app/components/ReactionsBar.vue]: Sorry forgot to link docs, it does work and IMHO is clearer to define classes as `class` if they don't need to be bind.

> You can have multiple classes toggled by having more fields in the object. In addition, the v-bind:class directive can also co-exist with the plain class attribute. 

https://vuejs.org/v2/guide/class-and-style.html#Object-Syntax

Comment (christophehenry) [reply socialhome/streams/app/components/ReactionsBar.vue]: > In addition, the `v-bind:class` directive can also co-exist with the plain `class` attribute.

Indeed.

File: socialhome/streams/app/components/Stream.vue (hunk 30 above)
Comment (jaywink) [reply socialhome/streams/app/components/Stream.vue:19]: OK, good enough for me! 👍

File: config/urls.py [hunk 31]
```diff
@@ -9,6 +9,7 @@
 from rest_framework.routers import DefaultRouter
 from rest_framework.schemas import get_schema_view
 from rest_framework_swagger.renderers import OpenAPIRenderer, SwaggerUIRenderer
+from django_js_reverse import views as django_js_reverse_views
```
Comment (jaywink) [inline config/urls.py]: I think you changed this since last review? Please don't use `as` imports unless there is no other way. It makes tracking annoying since the code calls something that isn't the real name what was imported. Even though PyCharm obviously helps, this is a very strong preference for me.

Always import the exact thing you need, unless there are lots of imports then I think it's ok to import a full module.

File: socialhome/static/sass/streams.scss [hunk 32]
```diff
@@ -12,21 +13,24 @@
 }
 
 @media (min-width: $bsMediumDevice) {
-  #new-content-container {
+  #new-content-container,
+  .new-content-container{
```
Comment (jaywink) [inline socialhome/static/sass/streams.scss]: Missing space before {.

File: socialhome/static/sass/streams.scss [hunk 33]
```diff
@@ -12,21 +13,24 @@
...
     font-size: 1.4em;
     top: 38px;
   }
 }
 
 @media (min-width: $bsLargeDevice) {
-  #new-content-container {
+  #new-content-container,
+  .new-content-container{
```
Comment (jaywink) [inline socialhome/static/sass/streams.scss]: Missing space before {.

File: socialhome/static/sass/streams.scss [hunk 34]
```diff
@@ -12,21 +13,24 @@
...
     font-size: 1.2em;
     top: 39px;
   }
 }
 
 @media (min-width: $bsXLargeDevice) {
-  #new-content-container {
+  #new-content-container,
+  .new-content-container{
```
Comment (jaywink) [inline socialhome/static/sass/streams.scss]: Missing space before {.

**Review by jaywink — CHANGES_REQUESTED** [review]
A few easy to fix things left - please don't change anything else to avoid more comments :D Then let's merge!

File: config/urls.py (hunk 31 above)
Comment (christophehenry) [reply config/urls.py]: > I think you changed this since last review?

No, I didn't. This `as` is quite old.

> Please don't use as imports unless there is no other way.

Even if what's imported as a very generic name?

**Comment by christophehenry** [comment]
Done.

File: config/urls.py [hunk 35]
```diff
@@ -9,6 +9,7 @@
 from rest_framework.routers import DefaultRouter
 from rest_framework.schemas import get_schema_view
 from rest_framework_swagger.renderers import OpenAPIRenderer, SwaggerUIRenderer
+from django_js_reverse import views
```
Comment (jaywink) [inline config/urls.py]: Importing views from a minor third party package is probably not a good idea. I meant import the thing you need only, so "from django_js_reverse.views import urls_js". If you look at other views files, direct import pattern is the regular way.

Comment (christophehenry) [reply config/urls.py]: Ah, right.

Comment (christophehenry) [reply config/urls.py]: Done

**Review by jaywink — APPROVED** [review]

**Comment by jaywink** [comment]
Thanks! <3
